server/src/main/java/io/crate/expression/scalar/DateBinFunction.java

```diff
@@ -0,0 +1,168 @@
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+import org.joda.time.Period;
+
+import java.util.List;
+
+public class DateBinFunction extends Scalar<Long, Object> {
+
+    public static final String NAME = "date_bin";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.INTERVAL.getTypeSignature(),
+                DataTypes.TIMESTAMPZ.getTypeSignature(), // source
+                DataTypes.TIMESTAMPZ.getTypeSignature(), // origin
+                DataTypes.TIMESTAMPZ.getTypeSignature()
+            ).withFeatures(DETERMINISTIC_ONLY),
+            DateBinFunction::new);
+
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.INTERVAL.getTypeSignature(),
+                DataTypes.TIMESTAMP.getTypeSignature(), // source
+                DataTypes.TIMESTAMP.getTypeSignature(), // origin
+                DataTypes.TIMESTAMP.getTypeSignature()
+            ).withFeatures(DETERMINISTIC_ONLY),
+            DateBinFunction::new);
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private DateBinFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public Scalar<Long, Object> compile(List<Symbol> arguments) {
+        assert arguments.size() == 3 : "Invalid number of arguments";
+
+        if (arguments.get(0) instanceof Input<?> input) {
+            var value = input.value();
+            if (value != null) {
+                long intervalInMs = ((Period) value).toStandardDuration().getMillis();
+                return new CompiledDateBin(intervalInMs);
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public final Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+        assert args.length == 3 : "Invalid number of arguments";
+
+        var interval = args[0].value();
+        var timestamp = args[1].value();
+        var origin = args[2].value();
+
+        if (interval == null || timestamp == null || origin == null) {
+            return null;
+        }
+
+        return getBinnedTimestamp(((Period) interval).toStandardDuration().getMillis(), (long) timestamp, (long) origin);
+    }
+
+    private static long getBinnedTimestamp(long interval, long timestamp, long origin) {
+        if (interval == 0) {
+            throw new IllegalArgumentException("Interval cannot be zero");
+        }
```

Comment:
Might be worth moving this check as well and do it only once for the compiled version.